### PR TITLE
feat: group dashboard tiles by tab

### PR DESCRIPTION
**Pull request by Shifan-Gu**
Closes: #9656
### Description:

ref: https://github.com/lightdash/lightdash/issues/9656

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

File: packages/backend/src/database/migrations/20240409160248_add_dashboard_tab.ts [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+            !(await knex.schema.hasColumn(
+                DashboardTilesTableName,
+                DashboardTileTabUuidColumnName,
+            ))
+        ) {
+            await knex.schema.alterTable(
+                DashboardTilesTableName,
+                (tableBuilder) => {
+                    tableBuilder
+                        .uuid(DashboardTileTabUuidColumnName)
```
Comment (ZeRego) [inline packages/backend/src/database/migrations/20240409160248_add_dashboard_tab.ts:64]: TODO: add reference to dashboard tabs table

File: packages/backend/src/database/migrations/20240409160248_add_dashboard_tab.ts [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+const DashboardTilesTableName = 'dashboard_tiles';
+const DashboardTileTabUuidColumnName = 'tab_uuid';
+
+type DbTabs = {
+    uuid: string;
+    name: string;
+    is_default: boolean;
+    dashboard_id: number;
+    dashboard_version_id: number;
+};
```
Comment (ZeRego) [inline packages/backend/src/database/migrations/20240409160248_add_dashboard_tab.ts:17]: TODO: delete unused type

File: packages/backend/src/database/migrations/20240409160248_add_dashboard_tab.ts [hunk 3]
```diff
@@ -0,0 +1,129 @@
+import { file } from 'googleapis/build/src/apis/file';
```
Comment (ZeRego) [inline packages/backend/src/database/migrations/20240409160248_add_dashboard_tab.ts:1]: TODO: remove unused import

File: packages/backend/src/models/DashboardModel/DashboardModel.ts [hunk 4]
```diff
@@ -571,6 +587,21 @@ export class DashboardModel {
                 dashboard.dashboard_version_id,
             );
 
+        const tabs = await this.database(DashboardTabsTableName)
+            .select<DashboardTab[]>(
+                `${DashboardTabsTableName}.name`,
+                `${DashboardTabsTableName}.uuid`,
+                `${DashboardTabsTableName}.dashboard_id`,
```
Comment (ZeRego) [inline packages/backend/src/models/DashboardModel/DashboardModel.ts:594]: TODO: we can remove `dashboard_id`

File: packages/frontend/src/pages/Dashboard.tsx [hunk 5]
```diff
@@ -375,13 +418,21 @@ const Dashboard: FC = () => {
 
     const handleAddTiles = useCallback(
         async (tiles: IDashboard['tiles'][number][]) => {
+            let newTiles = tiles;
+            if (activeTab?.uuid) {
+                newTiles = tiles.map((tile: DashboardTile) => ({
+                    ...tile,
+                    tabUuid: activeTab?.uuid,
+                }));
+                setHaveTabsChanged(true);
```
Comment (ZeRego) [inline packages/frontend/src/pages/Dashboard.tsx:427]: In this case, only the tiles properties/array changed (tile.tabUuid) and not the tabs.

File: packages/frontend/src/pages/Dashboard.tsx [hunk 6]
```diff
@@ -553,6 +612,50 @@ const Dashboard: FC = () => {
 
     const hasDashboardTiles = dashboardTiles && dashboardTiles.length > 0;
 
+    const currentTabHasTiles = sortedTiles?.some((tile) => {
+        return (
+            tile.tabUuid === activeTab?.uuid ||
+            (!tile.tabUuid && activeTab?.uuid === defaultTabUuid) ||
+            (tile.tabUuid && !tabs.find((t) => t.uuid === tile.tabUuid))
+        );
+    });
```
Comment (ZeRego) [inline packages/frontend/src/pages/Dashboard.tsx:621]: Found hard to follow what scenarios this covers. Would this logic work? 
```suggestion
    const currentTabHasTiles = sortedTiles?.some((tile) => {
       const tileTabUuid = tile.tableUuid ?? defaultTabUuid; // inherit default tab uuid if needed
        return  tileTabUuid === activeTab?.uuid;
    });
```

File: packages/frontend/src/pages/Dashboard.tsx [hunk 7]
```diff
@@ -553,6 +612,50 @@ const Dashboard: FC = () => {
...
+                    }
+                    return tab;
+                });
+                return newTabs;
+            });
+            setHaveTabsChanged(true);
+        }
+    };
+
+    const handleDeleteTab = (tabUuid: string) => {
```
Comment (ZeRego) [inline packages/frontend/src/pages/Dashboard.tsx:648]: TODO: remove the deleted tab uuid from tiles.

File: packages/backend/src/database/migrations/20240409160248_add_dashboard_tab.ts [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    dashboard_version_id: number;
+};
+
+export async function up(knex: Knex): Promise<void> {
+    // create dashboard tab table
+    if (!(await knex.schema.hasTable(tableName))) {
+        await knex.schema.createTable(tableName, (tableBuilder) => {
+            tableBuilder.string('name').notNullable();
+            tableBuilder.uuid('uuid').notNullable().defaultTo(uuidv4());
+            tableBuilder.boolean('is_default');
```
Comment (ZeRego) [inline packages/backend/src/database/migrations/20240409160248_add_dashboard_tab.ts:25]: This column doesn't seem to be used.

File: packages/backend/src/database/migrations/20240409160248_add_dashboard_tab.ts [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+                DashboardTilesTableName,
+                (tableBuilder) => {
+                    tableBuilder
+                        .uuid(DashboardTileTabUuidColumnName)
+                        .nullable();
+                },
+            );
+        }
+
+        // create a default tab for each dashboard
```
Comment (ZeRego) [inline packages/backend/src/database/migrations/20240409160248_add_dashboard_tab.ts:70]: We don't need to create a default tab. 

If dashboard has no tabs -> we render like we do now
If there are tabs -> we default tiles with no tab uuid to the first tab ( we need to track tab order somewhere )
If user deletes the last tab -> we render like we do now

File: packages/backend/src/database/migrations/20240409160248_add_dashboard_tab.ts (hunk 8 above)
Comment (ZeRego) [reply packages/backend/src/database/migrations/20240409160248_add_dashboard_tab.ts:16]: I still don't think the `is_default` is providing meaningful value. 🤔 
- If we delete the default tab there is no logic to set a new default 
- The code defaults to the first tab if it can't find the default one

If we remove this column we can rely only on the first tab available as being the default.

File: packages/frontend/src/components/DashboardTiles/TileForms/AddChartTilesModal.tsx [hunk 10]
```diff
@@ -100,6 +100,10 @@ const AddChartTilesModal: FC<Props> = ({ onAddTiles, onClose }) => {
                         chartName: chart?.name ?? '',
                     },
                     type: DashboardTileTypes.SAVED_CHART,
+                    tabUuid: (
+                        dashboard?.tabs.find((t) => t.isDefault) ??
+                        dashboard?.tabs?.[0]
+                    )?.uuid,
```
Comment (ZeRego) [inline packages/frontend/src/components/DashboardTiles/TileForms/AddChartTilesModal.tsx:106]: Not worth having this logic as we set the tabUuid in `handleAddTiles` (Dashboard.tsx)

```suggestion
                    tabUuid: undefined,
```

Comment (Shifan-Gu) [reply packages/frontend/src/components/DashboardTiles/TileForms/AddChartTilesModal.tsx:106]: will do 👍

File: packages/backend/src/database/migrations/20240409160248_add_dashboard_tab.ts (hunk 8 above)
Comment (ZeRego) [reply packages/backend/src/database/migrations/20240409160248_add_dashboard_tab.ts:16]: Based on the ticket requirements when a user deletes a tab we will delete all tiles inside it.

**Review by ZeRego — APPROVED** [review]
